### DISCUSSION TITLE
adding existing media files is still very slow.

### DISCUSSION THREAD

**Discussion by yob-yob** [spatie/laravel-medialibrary#2986] [Q&A]
So I'm creating an excel importer for my table `Deals`...

Each `Deals` has Many `Photoset`

and each `Photoset` has 3 media items associated with it... each media is under a single-file collection...

I'm currently using Azure-Blob-Storage for storing this images and is using the package `matthewbdaly/laravel-azure-storage` which is just a standard filesystem adapter for `league/flysystem-azure-blob-storage`

on my `Azure` Disk the all of the media-files already exists, all I need is to associate them into each photoset...

to do this I have this code... 

```php
foreach ($deals as $key => $deal) {
        foreach ($photosets[$deal->id] as $key => $photoset) {
          /** @var \App\Models\Photoset $set */
          $set = $deal->photosets()->create();
          $set->load('deal');
          static::handlePhoto($set, $photoset->{"14:9"}, 'original');
          static::handlePhoto($set, $photoset->{"1:1"}, 'square');
          static::handlePhoto($set, $photoset->{"2:1"}, 'landscape');
        }
}
```

`$deals` is all the deals, and `$photosets` is an array of...

```php
[
    {deal_id} => [
        "14:9" => filepath, 
        "1:1" => filepath, 
        "2:1" => filepath
    ],
    ...
]
```

the `handlePhoto` method is

```php
public static function handlePhoto(Photoset $set, $filepath, $collection)
{
      $set->addMediaFromDisk($filepath, 'azure')
             ->preservingOriginal()
             ->toMediaCollection($collection, 'azure');
}
```

So now I know what you're thinking... it's slow because media-library still needs to copy the image to a new folder...

but it doesn't really have to... 

here's why, I create a custom `PathGenerator` just for the photosets model... this is the new path structure...

```
$path = class_basename($media->model_type) . '/' . $media->collection_name . '/' . $media->model_id;
```

example output would be `Photoset/original/1/randomname.png`, I know you're probably concerned that there will be 2 media file in one directory.. but that's not the case, because as I've said, each media in a `photoset` is under a single-file collection.

here's the main problem...

as soon as `handlePhoto` is run it takes way too long to finish for each media... why? I don't really know... 😅

I did a little digging into the codebase and found that the code that handles this is on the method `toMediaCollectionFromRemote` 
under the `/spatie/laravel-medialibrary/src/MediaCollections/FileAdder.php` Class...

and I have check the steps it runs...

1. check if the file exists
2. gets the size of the file
3. gets the mimeType
4. gets the size of the file (Again)

each of those steps require a back-and-forth communication with azure which I assume is very resource consuming.

then after instantiating the media object, it then attach it to the model...

where `copyToMediaLibraryFromRemote` method is going to handle the `moving` of the file to the `destination`...
found on `spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php`
which is also another request to azure... and this method also show even more requests to azure....

it should not move the file anymore or copy it because it already exists in the same directory as where it wants it to be...

**Comment by freekmurze** [comment]
I agree that this should be optimised. 

Feel free to send a PR for this if you need this quickly.
